Sources/GRPC/Compression/MessageEncoding.swift

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright 2020, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+extension ClientConnection.Configuration {
+  public struct MessageEncoding {
+    public init(
+      requests: CompressionAlgorithm?,
+      responses: [CompressionAlgorithm]
+    ) {
+      self.outbound = requests
+      self.inbound = responses
+    }
+
+    /// The compression algorithm used for outbound messages.
+    public internal(set) var outbound: CompressionAlgorithm?
+
+    /// The set of compression algorithms advertised to the remote peer that they may use.
+    public internal(set) var inbound: [CompressionAlgorithm]
```

Comment:
Not super happy with these names, neither "requests/responses" nor "outbound/inbound". How about "accepted" and "outbound"? Or something more analogous to the "accept-encoding" headers? Not 100% sure myself. Thinking about this more, maybe the names are actually good.

Also, does gRPC support using different compressions in both directions on one stream? Or is this more like the "preferred" outbound method?

Overall, I'm a bit confused by this so far 😅

Reply:
Agh...naming here is tricky. To the user `requests`/`responses` seems more natural for the `init` than `inbound`/`outbound`, but then `responses` doesn't quite fit the "accept-encoding" part. Also `compression.requests`/`compression.responses` isn't as clear as `compression.outbound`, `compression.inbound`.

The `init` could be more explicit:

```swift
init(forRequests outbound: CompressionAlgorithm?, acceptableForResponses inbound: [CompressionAlgorithm]?) {
  // ...
}

...

MessageEncoding(forRequests: .gzip, acceptableForResponses: [.gzip, .identity])
```

I'm still not certain how this API should be spelt! Thoughts?

> Also, does gRPC support using different compressions in both directions on one stream? Or is this more like the "preferred" outbound method?

Compression for the request and response stream are independent and may be different.

Reply:
One other thing I've been considering recently which might make things easier for us and users is some sort of client bootstrap: we can be more descriptive with names and add to it without breaking API (with structs we need a new `init` if we add a property).

```swift
let bootstrap = GRPCConnectionBootstrap(group: group)  // hmm sounds familiar...
  .requestCompression(.gzip)
  .acceptableResponseCompression(.gzip, .deflate, .identity)
```

Anyway, that's a story for another time!

Reply:
`MessageEncoding(forRequests: .gzip, acceptableForResponses: [.gzip, .identity])` SGTM. I also like the Bootstrap idea.

Reply:
I'll leave the bootstrap for another day since that's a fairly large change.